### PR TITLE
Initial commit for discover-packet

### PR DESCRIPTION
**Pull request by jasmingacic**
Dependencies are not locked but I wasn't sure which dependency tool you were using. If someone can help me out with that.
Also is there anything else missing here?

### CODE REVIEW

**Comment by pearkes** [comment]
Hey there, this is looking like it is headed in the right direction, just have a few high level bits of feedback:

- Your vendoring should only include packages for Packet in this case. See [AWS](https://github.com/hashicorp/go-discover/blob/master/provider/aws/vendor/vendor.json) for an example. You did use the right tool here (govendor).
- It looks like you hardcode the test project. If possible, make this an environment variable so we can test in CI. We run automated tests for a lot of the providers, see an example with Google: https://github.com/hashicorp/go-discover/tree/master/test/tf/gce. If you can write some Terraform there we can set that up and make sure that the go-discover provider works end-to-end.
- Is a "project" in Packet similar to a "tag" in most other infrastructure providers? If you have the notion of "tags" it may be more suitable to use that.

If we can take care of those items I think this will otherwise be good!

**Comment by jasmingacic** [comment]
@pearkes I've just fixed dependencies and added project as environment variable. 

Listing Packet devices can be done only by querying `/projects/{id}/devices` I could fetch all projects and then iterate through each to fetch devices if that's what you would like to have here.
Also there is a parameter `with_flags` but it is not yet publicly accessible.

p.s. I will be adding TF files later tomorrow.

**Comment by jasmingacic** [comment]
@pearkes I've added TF tests.

File: provider/packet/packet_discover_test.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	p := packet.Provider{}
+
+	l := log.New(os.Stderr, "", log.LstdFlags)
+	addrs, err := p.Addrs(args, l)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(addrs) == 0 {
```
Comment (pearkes) [inline provider/packet/packet_discover_test.go]: Usually like to check the explicit number setup in the TF here (2 I think).

File: test/tf/packet/main.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
+provider "packet" {}
+
+resource "packet_project" "project" {
+  name = "go-discover:packet"
```
Comment (pearkes) [inline test/tf/packet/main.tf]: Can you add a bit of randomness here? See vsphere for [an example](https://github.com/hashicorp/go-discover/blob/master/test/tf/vsphere/main.tf#L89-L98). If this is a unique name in Packet that will help the tests from conflicting with each other.

File: test/tf/packet/main.tf [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+resource "packet_device" "discover-packet" {
+  count            = 2
+  hostname         = "tf.discover"
```
Comment (pearkes) [inline test/tf/packet/main.tf]: Randomness here if hostnames are unique (not sure).

Comment (jasmingacic) [reply test/tf/packet/main.tf]: Hostnames don't have to be unique, but i will randomize it anyways.

File: provider/packet/packet_discover.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	projectID := argsOrEnv(args, "packet_project", "PACKET_PROJECT")
+	packetURL := argsOrEnv(args, "packet_url", "PACKET_URL")
+	addressType := args["address_type"]
+
+	if addressType != "private_v4" && addressType != "public_v4" && addressType != "public_v6" {
+		l.Printf("[INFO] discover-packet: Address type %s is not supported. Valid values are {private_v4,public_v4,public_v6}. Falling back to 'private_v4'", addressType)
+		addressType = "private_v4"
+	}
+
+	if addressType == "" {
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply provider/packet/packet_discover.go]: i will remove it. It simply doesn't make sense to have it.

File: provider/packet/packet_discover_test.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	"testing"
+
+	discover "github.com/hashicorp/go-discover"
+	"github.com/hashicorp/go-discover/provider/packet"
+)
+
+var _ discover.Provider = (*packet.Provider)(nil)
+var _ discover.ProviderWithUserAgent = (*packet.Provider)(nil)
+
+func TestAddrs(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply provider/packet/packet_discover_test.go]: on it I will add tests for each existing type

**Comment by pearkes** [comment]
All looks good to me! This will be in a future version of Consul.

It'd be great to add docs here as well:

https://github.com/hashicorp/consul/blob/master/website/source/docs/agent/cloud-auto-join.html.md

**Pull request merged**